README.md
Embeddable-Badges-Plugin
==============================
[![Build Status](https://travis-ci.org/SxMShaDoW/embeddable-badges-plugin.svg?branch=master)](https://travis-ci.org/SxMShaDoW/embeddable-badges-plugin) [![codecov.io](https://codecov.io/github/SxMShaDoW/embeddable-badges-plugin/coverage.svg?branch=master)](https://codecov.io/github/SxMShaDoW/embeddable-badges-plugin?branch=master) [![Sputnik](https://sputnik.ci/conf/badge)](https://sputnik.ci/app#/builds/SxMShaDoW/embeddable-badges-plugin)

Inspired and based off of the the [Embeddable-Build-Status-Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Embeddable+Build+Status+Plugin), this project cleans up some of the URL types and adds in additional fun badges!

# Versions

See our [CHANGELOG](CHANGELOG.md) for a history of available versions and changes.

# Manual Installation

This plugin is not yet available for installation via Jenkins Plugin Manager (Manage Jenkins -> Manage Plugins), and requires manual installation.  The following steps are how to install it manually.

1. Sync the latest code line, or from one of our git tags.
2. Run `mvn clean package` locally, this generates a `target/embeddable-badges.hpi` file that you can manually install.
3. On the `Advanced` tab of the Jenkins plugin manager (`http://<your jenkins host>/pluginManager/advanced`), browse to the generated `target/embeddable-badges.hpi` file under the `Upload Plugin` section.
4. Click Upload.
5. Restart Jenkins.
6. Profit.

# Badge Types

### Build Status
Reports success or failure of the latest build.

### Code Coverage
Reports a coverage percentage for the following technologies:

1. Clover: Line coverage percentage
2. Cobertura: Element coverage percentage
3. Jacoco: Instruction coverage percentage

### Unit Tests
Number of JUnit or TestNG tests executed successfully.

### Description
With the [build description setter](https://wiki.jenkins-ci.org/display/JENKINS/Description+Setter+Plugin) plugin, folks often create regex's to post the version of the module that was created as a build description.  This badge will display whatever that description is.

# License

[MIT](LICENSE)


CHANGELOG.md
# Change Log

## next

* [rguenthe](https://github.com/rguenthe) Add [code coverage api](https://github.com/jenkinsci/code-coverage-api-plugin) line coverage support [#60](https://github.com/SxMShaDoW/embeddable-badges-plugin/pull/60)

## 2.5.2
[chetanddesai](https://github.com/chetanddesai)

* [YogiKhan](https://github.com/YogiKhan) Add view support [#54](https://github.com/SxMShaDoW/embeddable-badges-plugin/pull/54)
* [vijkris99](https://github.com/vijkris99) Fix badge truncation bug [#52](https://github.com/SxMShaDoW/embeddable-badges-plugin/pull/52)

## 2.5.1
[chetanddesai](https://github.com/chetanddesai)

* Fix broken Jenkins job URLs on Enterprise Jenkins [#45](https://github.com/SxMShaDoW/embeddable-badges-plugin/issues/45).


## 2.5
[chetanddesai](https://github.com/chetanddesai)

* First pass on dynamically scaling images [#40](https://github.com/SxMShaDoW/embeddable-badges-plugin/issues/40).

## 2.4
[chetanddesai](https://github.com/chetanddesai)

* Fix division cast issue on [#35](https://github.com/SxMShaDoW/embeddable-badges-plugin/issues/35).
* Add in a new build description badge [#34](https://github.com/SxMShaDoW/embeddable-badges-plugin/issues/34).

## 2.3
[chetanddesai](https://github.com/chetanddesai)

* Fix the colour bug on the test icon. See issue [#35](https://github.com/SxMShaDoW/embeddable-badges-plugin/issues/35).

## 2.2
[chetanddesai](https://github.com/chetanddesai)

* Subtract skipped tests from passed count, so if you skip `9` tests and `12` pass, it will read `12/21` instead of `21/21`.
* Scale `tests` image to support up to 9999 tests.
* Report `0` for no coverage results found instead of `na%`.

## 2.1
[chetanddesai](https://github.com/chetanddesai)

* Change coverage action data from `getBuildHealth().getScore()` to line percentage for Cobertura, and element coverage percentage for Clover.

## 2.0

* Lost in the ether. Who knows.

## 1.0

* Initial Release


src/test/java/org/jenkinsci/plugins/badge/PublicBadgeActionTest.java
/*
 * The MIT License
 *
 * Copyright 2013 Dominik Bartholdi.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.jenkinsci.plugins.badge;

import com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException;
import hudson.model.FreeStyleProject;
import hudson.security.GlobalMatrixAuthorizationStrategy;
import hudson.security.SecurityRealm;
import java.io.IOException;
import static java.lang.System.out;
import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
import static org.jenkinsci.plugins.badge.PublicBadgeAction.VIEW_STATUS;
import org.junit.After;
import org.junit.AfterClass;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.jvnet.hudson.test.JenkinsRule;
import org.jvnet.hudson.test.recipes.PresetData;
import static org.jvnet.hudson.test.recipes.PresetData.DataSet.ANONYMOUS_READONLY;
import static org.jvnet.hudson.test.recipes.PresetData.DataSet.NO_ANONYMOUS_READACCESS;
import org.kohsuke.stapler.StaplerRequest;
import org.kohsuke.stapler.StaplerResponse;

/**
 * @author Dominik Bartholdi (imod)
 */
@SuppressWarnings("unused")
public class PublicBadgeActionTest {

    /**
     *
     */
    @Rule
    public JenkinsRule j;

    /**
     *
     * @throws Exception
     */
    @BeforeClass
    public static void setUpClass() throws Exception {
    }

    /**
     *
     * @throws Exception
     */
    @AfterClass
    public static void tearDownClass() throws Exception {
    }
    /**
     *
     */
    public PublicBadgeActionTest() {
        this.j = new JenkinsRule();
    }

    /**
     *
     * @throws Exception
     */
    @Before
    public void setUp() throws Exception {
    }

    /**
     *
     * @throws Exception
     */
    @After
    public void tearDown() throws Exception {
    }

    /**
     *
     * @throws Exception
     */
    @PresetData(NO_ANONYMOUS_READACCESS)
    @Test
    public void authenticatedAccess() throws Exception {
        final FreeStyleProject project = j.createFreeStyleProject("free");
        JenkinsRule.WebClient wc = j.createWebClient();
        wc.login("alice", "alice");
        try {
            // try with wrong job name
            wc.goTo("buildStatus/buildIcon?job=dummy");
            fail("should fail, because there is no job with this name");
        } catch (FailingHttpStatusCodeException x) {
            assertEquals(HTTP_NOT_FOUND, x.getStatusCode());
        }
        wc.goTo("buildStatus/buildIcon?job=free", "image/svg+xml");
        j.buildAndAssertSuccess(project);
    }

    /**
     *
     * @throws Exception
     */
    @PresetData(NO_ANONYMOUS_READACCESS)
    @Test
    public void invalidAnonymousAccess() throws Exception {
        final FreeStyleProject project = j.createFreeStyleProject("free");
        JenkinsRule.WebClient wc = j.createWebClient();
        try {
            // try with wrong job name
            wc.goTo("buildStatus/buildIcon?job=dummy");
            fail("should fail, because there is no job with this name");
        } catch (FailingHttpStatusCodeException x) {
            assertEquals(HTTP_NOT_FOUND, x.getStatusCode());
        }

        try {
            // try with correct job name
            wc.goTo("buildStatus/buildIcon?job=free", "image/svg+xml");
            fail("should fail, because there is no job with this name");
        } catch (FailingHttpStatusCodeException x) {
            // make sure return code does not leak security relevant information (must 404)
            assertEquals(HTTP_NOT_FOUND, x.getStatusCode());
        }

        j.buildAndAssertSuccess(project);

    }

    /**
     *
     * @throws Exception
     */
    @Test
    public void validAnonymousViewStatusAccess() throws Exception {

        final SecurityRealm realm = j.createDummySecurityRealm();
        j.jenkins.setSecurityRealm(realm);
        GlobalMatrixAuthorizationStrategy auth = new GlobalMatrixAuthorizationStrategy();
        auth.add(VIEW_STATUS, "anonymous");
        j.getInstance().setSecurityRealm(realm);
        j.getInstance().setAuthorizationStrategy(auth);

        final FreeStyleProject project = j.createFreeStyleProject("free");

        JenkinsRule.WebClient wc = j.createWebClient();
        try {
            // try with wrong job name
            wc.goTo("buildStatus/buildIcon?job=dummy");
            fail("should fail, because there is no job with this name");
        } catch (FailingHttpStatusCodeException x) {
            assertEquals(HTTP_NOT_FOUND, x.getStatusCode());
        }

        wc.goTo("buildStatus/buildIcon?job=free", "image/svg+xml");
        j.buildAndAssertSuccess(project);
   
    }
    
    /**
     *
     * @throws Exception
     */
    @PresetData(ANONYMOUS_READONLY)
    @Test
    public void validAnonymousAccess() throws Exception {
        final FreeStyleProject project = j.createFreeStyleProject("free");
        JenkinsRule.WebClient wc = j.createWebClient();
        try {
            // try with wrong job name
            wc.goTo("buildStatus/buildIcon?job=dummy");
            fail("should fail, because there is no job with this name");
        } catch (FailingHttpStatusCodeException x) {
            assertEquals(HTTP_NOT_FOUND, x.getStatusCode());
        }

        // try with correct job name
        wc.goTo("buildStatus/buildIcon?job=free", "image/svg+xml");
        j.buildAndAssertSuccess(project);
    }

    /**
     * Test of getUrlName method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
    @Test
    public void testGetUrlName() throws IOException {
        out.println("getUrlName");
        PublicBadgeAction instance = new PublicBadgeAction();
        String expResult = "buildStatus";
        String result = instance.getUrlName();
        assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }

    /**
     * Test of getIconFileName method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
    @Test
    public void testGetIconFileName() throws IOException {
        out.println("getIconFileName");
        PublicBadgeAction instance = new PublicBadgeAction();
        String expResult = null; //Change this eventually
        String result = instance.getIconFileName();
        assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }

    /**
     * Test of getDisplayName method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
    @Test
    public void testGetDisplayName() throws IOException {
        out.println("getDisplayName");
        PublicBadgeAction instance = new PublicBadgeAction();
        String expResult = null; //change this eventually
        String result = instance.getDisplayName();
        assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }

    /**
     * Test of doCoverageIcon method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
	@Test
    public void testDoCoverageIcon() throws IOException {
        out.println("doCoverageIcon");
        StaplerRequest req = null;
        StaplerResponse rsp = null;
        String job = "";
        PublicBadgeAction instance = new PublicBadgeAction();
        //HttpResponse expResult = instance.doCoverageIcon(req, rsp, job);
        //HttpResponse result = instance.doCoverageIcon(req, rsp, job);
        assertEquals(1, 1);
        //assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }

    /**
     * Test of doTestIcon method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
    @Test
    public void testDoTestIcon() throws IOException {
        out.println("doTestIcon");
        StaplerRequest req = null;
        StaplerResponse rsp = null;
        String job = "";
        PublicBadgeAction instance = new PublicBadgeAction();
        //HttpResponse expResult = instance.doTestIcon(req, rsp, job);
        //HttpResponse result = instance.doTestIcon(req, rsp, job);
        assertEquals(1, 1);
        //assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }

    /**
     * Test of doBuildIcon method, of class PublicBadgeAction.
     * @throws java.io.IOException
     */
    @Test
    public void testDoBuildIcon() throws IOException {
        out.println("doBuildIcon");
        StaplerRequest req = null;
        StaplerResponse rsp = null;
        String job = "";
        PublicBadgeAction instance = new PublicBadgeAction();
        //HttpResponse expResult = instance.doBuildIcon(req, rsp, job);
        //HttpResponse result = instance.doBuildIcon(req, rsp, job);
        assertEquals(1, 1);
        //assertEquals(expResult, result);
        // TODO review the generated test code and remove the default call to fail.
        //fail("The test case is a prototype.");
    }
}

src/test/java/org/jenkinsci/plugins/badge/WorkflowBadgeActionTest.java
package org.jenkinsci.plugins.badge;


import hudson.security.GlobalMatrixAuthorizationStrategy;
import hudson.security.SecurityRealm;
import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
import org.jenkinsci.plugins.workflow.job.WorkflowJob;
import org.junit.Rule;
import org.junit.Test;
import org.jvnet.hudson.test.JenkinsRule;
import org.jvnet.hudson.test.recipes.PresetData;


/**
 * Just the fun bits: check that actions register correctly
 */
public class WorkflowBadgeActionTest {

    @Rule
    public JenkinsRule j = new JenkinsRule();

    @PresetData(PresetData.DataSet.NO_ANONYMOUS_READACCESS)
    @Test
    public void authenticatedAccess() throws Exception {
        WorkflowJob job = j.getInstance().createProject(WorkflowJob.class, "wf");
        job.setDefinition(new CpsFlowDefinition("println('hello')"));

        JenkinsRule.WebClient wc = j.createWebClient();
        wc.login("alice", "alice");

        wc.goTo("buildStatus/buildIcon?job=wf", "image/svg+xml");
        job.setQuietPeriod(0);
        job.scheduleBuild();
        j.waitUntilNoActivityUpTo(5000);
        wc.goTo("buildStatus/buildIcon?job=wf", "image/svg+xml");
    }

    @Test
    public void anonymousViewStatusAccess() throws Exception {
        // Allows anonymous access at security realm level
        final SecurityRealm realm = j.createDummySecurityRealm();
        j.jenkins.setSecurityRealm(realm);
        GlobalMatrixAuthorizationStrategy auth = new GlobalMatrixAuthorizationStrategy();
        auth.add(PublicBadgeAction.VIEW_STATUS, "anonymous");
        j.getInstance().setSecurityRealm(realm);
        j.getInstance().setAuthorizationStrategy(auth);

        WorkflowJob job = j.getInstance().createProject(WorkflowJob.class, "wf");
        job.setDefinition(new CpsFlowDefinition("println('hello')"));
        JenkinsRule.WebClient wc = j.createWebClient();
        wc.goTo("buildStatus/buildIcon?job=wf", "image/svg+xml");
    }

    @PresetData(PresetData.DataSet.ANONYMOUS_READONLY)
    @Test
    public void anonymousRead() throws Exception {
        WorkflowJob job = j.getInstance().createProject(WorkflowJob.class, "wf");
        job.setDefinition(new CpsFlowDefinition("println('hello')"));
        JenkinsRule.WebClient wc = j.createWebClient();
        wc.goTo("buildStatus/buildIcon?job=wf", "image/svg+xml");
    }
}


src/main/java/org/jenkinsci/plugins/badge/PluginImpl.java
/*
* The MIT License
*
* Copyright 2013 Dominik Bartholdi.
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/
package org.jenkinsci.plugins.badge;

import hudson.Plugin;
import static org.jenkinsci.plugins.badge.PublicBadgeAction.VIEW_STATUS;

/**
 * This plugin implementation only exists to force the loading of the permission
 * in an early enough stage (see also JENKINS-4172).
 * If the permission is not loaded early enough, Jenkins fails to load
 * permissions from config.
 *
 * @author Dominik Bartholdi (imod)
 */
public class PluginImpl extends Plugin {

    /**
     * TO DO
     * @throws Exception
     */
    @Override
    public void start() throws Exception {
        //
        // As a work around, force loading of this permission so that by
        // the time we start loading ACLs,
        // we have this instance already registered, thereby avoiding a lookup.
        VIEW_STATUS.toString();
    }

}


src/main/java/org/jenkinsci/plugins/badge/RunBadgeAction.java
package org.jenkinsci.plugins.badge;

import hudson.model.Action;
import hudson.model.Job;
import hudson.model.Run;
import jenkins.model.Jenkins;
import org.kohsuke.stapler.HttpResponse;
import org.kohsuke.stapler.QueryParameter;

public class RunBadgeAction implements Action {
    private final RunBadgeActionFactory factory;
    public final Run run;
    public final Job project;

    public RunBadgeAction(RunBadgeActionFactory factory, Run run) {
        this.factory = factory;
        this.run = run;
        this.project = run.getParent();
    }

    public String getIconFileName() {
        return Jenkins.RESOURCE_PATH+"/plugin/embeddable-badges/images/24x24/shield.png";
    }

    public String getDisplayName() {
        return Messages.RunBadgeAction_DisplayName();
    }

    public String getUrlName() {
        return "badge";
    }

    /**
     * Serves the badge image.
     */
    public HttpResponse doIcon(@QueryParameter String style) {
        return factory.getImage(run.getIconColor(), style);
    }
}


src/main/java/org/jenkinsci/plugins/badge/ImageResolver.java
/*
* The MIT License
*
* Copyright 2013 Kohsuke Kawaguchi, Dominik Bartholdi
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
 */
package org.jenkinsci.plugins.badge;

import hudson.model.BallColor;
import hudson.model.HealthReport;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;

import static jenkins.model.Jenkins.getInstance;
import org.apache.commons.io.IOUtils;
import static org.apache.commons.io.IOUtils.toInputStream;

/**
 * TO DO
 * @author dkersch
 */
public class ImageResolver {
    /**
     * TO DO
     */
    private final HashMap<String, StatusImage[]> styles;
    /**
     * TO DO
     */
    private final StatusImage[] defaultStyle;
    /**
     * TO DO
     */
    public static final String RED = "#cc0000";
    /**
     * TO DO
     */
    public static final String YELLOW = "#b2b200";
    /**
     * TO DO
     */
    public static final String GREEN = "#008000";
    /**
     * TO DO
     */
    public static final String GREY = "#808080";
    /**
     * TO DO
     */
    public static final String BLUE = "#007ec6";
    
    // Include the plugin's current version number in the ETag,...
	// ...so that cached responses from a previous version don't get served
    private final String thisPluginCurrentVersion = getClass().getPackage().getImplementationVersion();

    
    /**
     * TO DO
     * @throws IOException
     */
    public ImageResolver() throws IOException {
        styles = new HashMap<String, StatusImage[]>();
        // shields.io "flat" style (new default from Feb 1 2015)
        StatusImage[] flatImages;
        flatImages = new StatusImage[] {
            new StatusImage("build-failing-red-flat.svg"),
            new StatusImage("build-unstable-yellow-flat.svg"),
            new StatusImage("build-passing-brightgreen-flat.svg"),
            new StatusImage("build-running-blue-red-flat.svg"),
            new StatusImage("build-running-blue-yellow-flat.svg"),
            new StatusImage("build-running-blue-brightgreen-flat.svg"),
            new StatusImage("build-running-blue-lightgrey-aborted-flat.svg"),
            new StatusImage("build-running-blue-lightgrey-unknown-flat.svg"),
            new StatusImage("build-aborted-lightgrey-flat.svg"),
            new StatusImage("build-unknown-lightgrey-flat.svg")
        };
        defaultStyle = flatImages;
        styles.put("default", defaultStyle);
    }

    /**
     * TO DO
     * @param codeCoverage
     * @return
     */
    public StatusImage getCoverageImage(Integer codeCoverage) {
        // TODO don't read file everytime, store this as a static variable in
        // TODO memory with the constructor
        URL image = null;
        try {
            image = new URL(
                    getInstance().pluginManager.getPlugin("embeddable-badges").baseResourceURL,
                    "status/build-coverage-flat.svg");
        } catch (MalformedURLException e1) {
            // TODO Auto-generated catch block
            e1.printStackTrace();
        }

        StringBuilder sb = null;
        try {
            sb = new StringBuilder(IOUtils.toString(image.openStream()));
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        String replacedImage = replaceCodeCoverageSVG(sb.toString(), codeCoverage);
        InputStream is = toInputStream(replacedImage);
        String etag = thisPluginCurrentVersion + "-status/build-coverage-flat.svg-" + codeCoverage;

        try {
            return new StatusImage(etag, is);
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        
        return null;
    }
    
    /**
     * TO DO
     * @param image
     * @param codeCoverage
     * @return
     */
    private String replaceCodeCoverageSVG(String image, Integer codeCoverage) {
        if (codeCoverage == null) {
            String modifiedColor = image.replace("{hex-color-to-change}", GREY);
            //return modifiedColor.replace("{code-coverage-to-change}", "0");
            return modifiedColor.replace("{code-coverage-to-change}", "NaN");

        } else if (codeCoverage < 20) {
            String modifiedColor = image.replace("{hex-color-to-change}", RED);
            return modifiedColor.replace("{code-coverage-to-change}", codeCoverage.toString());
        } else if (codeCoverage < 80) {
            String modifiedColor = image.replace("{hex-color-to-change}", YELLOW);
            return modifiedColor.replace("{code-coverage-to-change}", codeCoverage.toString());
        } else {
            String modifiedColor = image.replace("{hex-color-to-change}", GREEN);
            return modifiedColor.replace("{code-coverage-to-change}", codeCoverage.toString());
        }
    }
    
    private Integer[] scaleBadge(Integer totalLength, String badgeType){
        //52 (x6), 70 (x12) and 82 (x3), 95 (x6)
        int defaultTextLength = 52; //default for test status. Old: 52
        int defaultBadgeLength = 70; //default for test status. Old: 70
        int defaultMultiplier = 3;	// Old: 6
        
        if ("description".equals(badgeType)) {
            defaultTextLength = 82;	// Old: 82
            defaultBadgeLength = 95;	// Old: 95
            defaultMultiplier = 3;	// Old: 3
        }
        
        if (totalLength > 3) {	// > "n/a"
            defaultTextLength = defaultTextLength + (defaultMultiplier * totalLength);
            defaultBadgeLength = defaultBadgeLength + ((defaultMultiplier * 2) * totalLength);
        }
        
        Integer[] svgLengths = new Integer[2];
        svgLengths[0] = defaultTextLength;
        svgLengths[1] = defaultBadgeLength;
        
        return svgLengths;
    }
    
    /**
     * TO DO
     * @param testPass
     * @param testTotal
     * @return
     */
    public StatusImage getTestResultImage(Integer testPass, Integer testTotal) {
        // TODO don't read file everytime
        // TODO store this as a static variable in memory with the constructor
        URL image = null;
        try {
            image = new URL(
                    getInstance().pluginManager.getPlugin("embeddable-badges").baseResourceURL,
                    "status/build-test-result-flat.svg");
        } catch (MalformedURLException e1) {
            // TODO Auto-generated catch block
            e1.printStackTrace();
        }
        
        StringBuilder sb = null;
        try {
            sb = new StringBuilder(IOUtils.toString(image.openStream()));
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        String replacedImage = replaceTestResultSVG(sb.toString(), testPass, testTotal);
        InputStream is = toInputStream(replacedImage);
        String etag = thisPluginCurrentVersion + "-status/build-test-result-flat.svg-" + testPass + testTotal;

        try {
            return new StatusImage(etag, is);
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        
        return null;
    }
    
    /**
     * TO DO
     * @param image
     * @param testPass
     * @param testTotal
     * @return
     */
    private String replaceTestResultSVG(String image, Integer testPass, Integer testTotal) {
        int totalLength;
        String badgeType;
        badgeType = "tests";
        try {
        	totalLength = testPass.toString().length() + testTotal.toString().length() + 1; //include the "/" in the badge
        } catch (Exception ex) {
        	totalLength = 3;	// "n/a"
        }
        
        Integer[] svgLengths = scaleBadge(totalLength, badgeType);
        Integer defaultTextLength = svgLengths[0];
        Integer defaultBadgeLength = svgLengths[1];
        String modifiedTextLength = image.replace("{change-text-length}", String.valueOf(defaultTextLength));
        String modifiedBadgeLength = modifiedTextLength.replace("{change-badge-length}", String.valueOf(defaultBadgeLength));
        
        if (testTotal == null || testPass == null) {
            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", GREY);
            return modifiedColor.replace("{passed-tests}/{total-tests}", "n/a");
        } 
        else {
                double passTest = (double) (testPass);
                double passTotal = (double) (testTotal);
        	double passPercent = (passTest / passTotal) * 100.0;
        	if (passPercent < 20) {
	            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", RED);
	            String modifiedPass = modifiedColor.replace("{passed-tests}", testPass.toString());
	            return modifiedPass.replace("{total-tests}", testTotal.toString());
	        } else if (passPercent < 80) {
	            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", YELLOW);
	            String modifiedPass = modifiedColor.replace("{passed-tests}", testPass.toString());
	            return modifiedPass.replace("{total-tests}", testTotal.toString());
	        } else {
	            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", GREEN);
	            String modifiedPass = modifiedColor.replace("{passed-tests}", testPass.toString());
	            return modifiedPass.replace("{total-tests}", testTotal.toString());
	        }
        }
    }
    
    /**
     * TO DO
     * @param color
     * @return
     */
    public StatusImage getImage(BallColor color, BallColor lastBuildColor) {
        StatusImage[] images = styles.get("default");

        if (color.isAnimated()) {	// Running
        	return getRunningBuildImage(lastBuildColor);
        }

        switch (color) {
            case RED:
                return images[0];
            case YELLOW:
                return images[1];
            case BLUE:
                return images[2];
            case ABORTED:
                return images[8];
            default:
                return images[9];
        }
    }
    
    private StatusImage getRunningBuildImage(BallColor lastBuildColor) {
    	StatusImage[] images = styles.get("default");

        switch (lastBuildColor) {
            case RED:
                return images[3];
            case YELLOW:
                return images[4];
            case BLUE:
                return images[5];
            case ABORTED:
                return images[6];
            default:
                return images[7];
        }
    }
    
    /**
     * TO DO
     * @param image
     * @param testPass
     * @param testTotal
     * @return
     */
    private String replaceBuildDescriptionSVG(String image, String buildDescription) {
        int totalLength;
        String badgeType;
        badgeType = "description";
        try {
        	totalLength = buildDescription.length();
        } catch (Exception ex) {
        	totalLength = 3;	// "n/a"
        }
        
        Integer[] svgLengths = scaleBadge(totalLength, badgeType);
        Integer defaultTextLength = svgLengths[0];
        Integer defaultBadgeLength = svgLengths[1];
        
        String modifiedTextLength = image.replace("{change-text-length}", String.valueOf(defaultTextLength));
        String modifiedBadgeLength = modifiedTextLength.replace("{change-badge-length}", String.valueOf(defaultBadgeLength));
        
        if (buildDescription == null) {
            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", GREY);
            return modifiedColor.replace("{description_text}", "n/a");
        }
        else {
            String modifiedColor = modifiedBadgeLength.replace("{hex-color-to-change}", BLUE);
            String modifiedPass = modifiedColor.replace("{description_text}", buildDescription);
            return modifiedPass;
        }
    }
    
    public StatusImage getBuildDescriptionImage(String buildDescription) {
        // TODO don't read file everytime
        // TODO store this as a static variable in memory with the constructor
        URL imageUrl = null;
        try {
            imageUrl = new URL(
                    getInstance().pluginManager.getPlugin("embeddable-badges").baseResourceURL,
                    "status/build-description-flat.svg");
        } catch (MalformedURLException e1) {
            // TODO Auto-generated catch block
            e1.printStackTrace();
        }
        StringBuilder sb = null;
        try {
            sb = new StringBuilder(IOUtils.toString(imageUrl.openStream()));
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        String replacedImage = replaceBuildDescriptionSVG(sb.toString(), buildDescription);
        InputStream is = toInputStream(replacedImage);
        String etag = thisPluginCurrentVersion + "-status/build-description-flat.svg-" + buildDescription;

        try {
            return new StatusImage(etag, is);
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return null;
    }
    
    public StatusImage getWeatherImage(HealthReport healthReport) {
    	// TODO don't read file everytime
        // TODO store this as a static variable in memory with the constructor
        URL imageUrl = null;
        String imageName;
        
        int score = healthReport.getScore();
        if (score <= 20) {
            imageName = "weather/health-00to19.svg";
        } else if (score <= 40) {
        	imageName = "weather/health-20to39.svg";
        } else if (score <= 60) {
        	imageName = "weather/health-40to59.svg";
        } else if (score <= 80) {
        	imageName = "weather/health-60to79.svg";
        } else {
        	imageName = "weather/health-80plus.svg";
        }
        
        try {
            imageUrl = new URL(
                    getInstance().pluginManager.getPlugin("embeddable-badges").baseResourceURL,
                    imageName);
        } catch (MalformedURLException e1) {
            // TODO Auto-generated catch block
            e1.printStackTrace();
        }
    	
        StringBuilder sb = null;
        try {
            sb = new StringBuilder(IOUtils.toString(imageUrl.openStream()));
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        InputStream is = toInputStream(sb.toString());
        String etag = thisPluginCurrentVersion + "-" + imageName;

        try {
            return new StatusImage(etag, is);
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        
    	return null;
    }
}